### PR TITLE
ORC-1205: `nextVector` should invoke `ensureSize` when reusing vectors

### PR DESCRIPTION
**Pull request by abstractdog**
### What changes were proposed in this pull request?
The fix simply ensures the size of some batches in ConvertTreeReader instances without API change.

### Why are the changes needed?
As described on jira, if there are batches of increasing size while reading, if they are in different stripes, we can hit an edge case where nextBatch doesn't force them to ensure the size of some batches.

### How was this patch tested?
Unit tests added for all possible ConvertTreeReader, also tested in hive locally.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
BTW, @abstractdog . Which ORC version are you using now?
- The latest one is https://orc.apache.org/news/2022/06/16/ORC-1.7.5/ 
- 1.6.x will be EOL soon.

**Comment by guiyanakuang** [comment]
Thanks to @abstractdog  for the extensive test coverage. I verified that the patch works.

The current reader inheritance relationship looks like this

```
TypeReader +--> TreeReader +---> ConvertTreeReader +-> Various ConvertTreeReaderImpl
                    +
                    |
                    +--------------------------------> Various TreeReaderImpl
```
The new ensureSize interface is called recursively to ensure the size of the vector, which is a good solution to the out-of-bounds problem in Various ConvertTreeReaderImpl. However, in the Various TreeReaderImpl code path we had the ensureSize method being called repeatedly. I think this is a good refactoring design and I wonder if we should simply fix the out-of-bounds issue in the ConvertTreeReader first. Then refactor the ensureSize entry in another pr to be consistent.

Please let me know if I am not correct.

**Comment by abstractdog** [comment]
thanks for the comments so far!

@dongjoon-hyun: hive is on ORC 1.6.9, so I'm motivated to backport the patch to branch-1.6 too if it can make its way to at least a single release, but if it becomes EOL, I don't mind, we're fixing this on 1.7.x too, hive should upgrade eventually anyway :)

@guiyanakuang: can you please elaborate on this:
```
in the Various TreeReaderImpl code path we had the ensureSize method being called repeatedly
```
it's not clear where exactly this happens, and why it is a problem, but otherwise, I'm fine with fixing this on ConvertTreeReaders only: so did you mean that the ensureSize method can stay exposed in TypeReader interface (as it is in the current patch), and let it be a no-op in TreeReaderFactory (so ConvertTreeReader will implement it instead of calling super.ensureSize)?

**Comment by guiyanakuang** [comment]
What I mean is that this pr is more like a refactoring than a problem fix.

The ensureSize interface is defined in the TypeReader, and if this pr is treated as a refactoring, our modification looks only half done. There are still many temporary reuse vector calls to ensureSize in Various TreeReaderImpl, rather than using the new interface to control size (e.g. line 2131).

https://github.com/apache/orc/blob/3eac56761f48f3e315eea56dc63de9f2cbadcdbb/java/core/src/java/org/apache/orc/impl/TreeReaderFactory.java#L2126-L2134

So I think it would be better to use a separate pr to abstract the code and use the public ensureSize interface (both ConvertTreeReaderImpl and TreeReaderImpl use this interface to control the size of the vectors).

**Comment by guiyanakuang** [comment]
> I see, makes sense @guiyanakuang as I'm now interested in fixing the issue ASAP for ConvertTreeReaders, the best way is what you proposed in Jira, right?
> 
> ```
>     if (doubleColVector == null) {
>         // Allocate column vector for file; cast column vector for reader.
>         doubleColVector = new DoubleColumnVector(batchSize);
>         longColVector = (LongColumnVector) previousVector;
>     }
> +   else {
> +       doubleColVector.ensureSize(batchSize, false);
> +   }
> ```
> 
> without touching the interface
> 
> if so, let me refresh the PR soon with the fix

Yes I do think so.

**Comment by abstractdog** [comment]
thanks @guiyanakuang, just force pushed the fix

File: java/core/src/java/org/apache/orc/impl/ConvertTreeReaderFactory.java [hunk 1]
```diff
@@ -1764,6 +1812,9 @@ public void nextVector(ColumnVector previousVector,
                                                    " proleptic Gregorian dates.");
           }
         }
+      } else {
+        bytesColVector.ensureSize(batchSize, false);
+        dateColumnVector.ensureSize(batchSize, false);
```
Comment (guiyanakuang) [inline java/core/src/java/org/apache/orc/impl/ConvertTreeReaderFactory.java:1817]: `dateColumnVector.ensureSize(batchSize, false);` may throw a null pointer exception and we may need to check additionally.

Comment (guiyanakuang) [reply java/core/src/java/org/apache/orc/impl/ConvertTreeReaderFactory.java:1817]: `dateColumnVector` can be ignored. This is because it is the vector obtained by forcing `previousVector` .

https://github.com/apache/orc/blob/3eac56761f48f3e315eea56dc63de9f2cbadcdbb/java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java#L1423

`computeBatchSize` ensures that batchSize <= the size of the input vector.

Comment (abstractdog) [reply java/core/src/java/org/apache/orc/impl/ConvertTreeReaderFactory.java]: handled it, thanks!

Comment (guiyanakuang) [reply java/core/src/java/org/apache/orc/impl/ConvertTreeReaderFactory.java]: > `dateColumnVector` can be ignored. This is because it is the vector obtained by forcing `previousVector` .
> 
> https://github.com/apache/orc/blob/3eac56761f48f3e315eea56dc63de9f2cbadcdbb/java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java#L1423
> 
> `computeBatchSize` ensures that batchSize <= the size of the input vector.

@abstractdog, in addition to this, LGTM

Comment (abstractdog) [reply java/core/src/java/org/apache/orc/impl/ConvertTreeReaderFactory.java]: okay, removed
thanks a lot for the quick reviews @guiyanakuang !

**Comment by abstractdog** [comment]
thanks for your comments and approving this PR @guiyanakuang 
if patch looks good, I'm about to open some PR with the very same contents, I'll update jira

**Comment by dongjoon-hyun** [comment]
In addition, 
- Please revise the PR description. The current one is invalid.
> This fix introduces a new API for TreeReaders: ensureSize

- After revising the PR description, please copy and paste it into your new PRs. PR title and description become a commit title and message. So, the following is an invalid PR description in your backporting PRs in Apache ORC community.
> same patch as main

**Comment by abstractdog** [comment]
finally, this doesn't introduce any APIs, updating descriptions

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you so much, @abstractdog and @guiyanakuang !
Merged to main for Apache ORC 1.9.0.
